databricks/koalas/frame.py

```diff
@@ -15,55 +15,65 @@
 #
 
 """
-A base class to be monkey-patched to DataFrame to behave similar to pandas DataFrame.
+A wrapper class for Spark DataFrame to behave similar to pandas DataFrame.
 """
+from decorator import dispatch_on
 from functools import reduce
 
 import numpy as np
 import pandas as pd
-from pyspark.sql import Column, DataFrame, functions as F
+from pyspark import sql as spark
+from pyspark.sql import functions as F
 from pyspark.sql.types import StructType, to_arrow_type
 from pyspark.sql.utils import AnalysisException
 
 from databricks.koalas.dask.compatibility import string_types
 from databricks.koalas.dask.utils import derived_from
-from databricks.koalas.generic import _Frame, anchor_wrap, max_display_count
+from databricks.koalas.generic import _Frame, max_display_count
 from databricks.koalas.metadata import Metadata
 from databricks.koalas.missing.frame import _MissingPandasLikeDataFrame
 from databricks.koalas.selection import SparkDataFrameLocator
 
 
-class PandasLikeDataFrame(_Frame, _MissingPandasLikeDataFrame):
-    """
-    Methods that are relevant to dataframes.
-    """
+def default_session():
+    return spark.SparkSession.builder.getOrCreate()
 
-    def __init__(self):
-        """ Define additional private fields.
 
-        * ``_pandas_metadata``: The metadata which stores column fields, and index fields and names.
-        """
-        self._pandas_metadata = None
+class DataFrame(_Frame, _MissingPandasLikeDataFrame):
+
+    @derived_from(pd.DataFrame)
+    @dispatch_on('data')
+    def __init__(self, data=None, index=None, columns=None, dtype=None, copy=False):
+        pdf = pd.DataFrame(data=data, index=index, columns=columns, dtype=dtype, copy=copy)
+        self._init_from_pandas(pdf)
+
+    @__init__.register(pd.DataFrame)
+    def _init_from_pandas(self, pdf, *args):
+        metadata = Metadata.from_pandas(pdf)
+        reset_index = pdf.reset_index()
+        reset_index.columns = metadata.all_fields
+        self._init_from_spark(default_session().createDataFrame(reset_index), metadata)
+
+    @__init__.register(spark.DataFrame)
+    def _init_from_spark(self, sdf, metadata=None, *args):
+        self._sdf = sdf
+        self._pandas_metadata = metadata
```

Comment:
How about .. 

```diff
diff --git a/databricks/koalas/frame.py b/databricks/koalas/frame.py
index b5fb148..2e1bd65 100644
--- a/databricks/koalas/frame.py
+++ b/databricks/koalas/frame.py
@@ -57,12 +57,13 @@ class DataFrame(_Frame, _MissingPandasLikeDataFrame):
     @__init__.register(spark.DataFrame)
     def _init_from_spark(self, sdf, metadata=None, *args):
         self._sdf = sdf
-        self._pandas_metadata = metadata
+        if metadata is None:
+            self._pandas_metadata = Metadata(column_fields=self._sdf.schema.fieldNames())
+        else:
+            self._pandas_metadata = metadata

     @property
     def _metadata(self):
-        if self._pandas_metadata is None:
-            self._pandas_metadata = Metadata(column_fields=self._sdf.schema.fieldNames())
         return self._pandas_metadata

     @property
```

We could even remove `_metadata` by renaming `_pandas_metadata` to that but I guess we can do this cleanup separately too.

Seems like it's a bug if `_pandas_metadata` becomes `None` somehow.